basic boolean conversion and interpretation behavior

kellyredding opened the pull request:
The idea here is to interpret incoming values as booleans in a friendly,
flexible way and be able to easily compare them to other booleans (both
`MuchBoolean` and `TrueClass` and `FalseClass`).  Use `convert` to just
convert a given value to an actual boolean value using much-boolean's
"friendly" interpretation logic.  Create instances using `new` to track
a given values interpreted boolean value and use as a boolean in
comparisons, etc.

Note: this was extracted from the ns-options gem with a few differences:
- no `actual=` writer (this was only needed b/c of ns-options behavior)
- no `returned_value` method (this was only needed b/c of ns-options
  behavior)
- support for "yes/no" type values
- support for additional "true/false" type values
- exposed the underlying conversion logic in a `convert` singleton so
  you don't _have_ to build an instance and get its `actual` value
  if all you care about is the actual boolean value

@jcredding ready for review.  Are you cool with the departures I did from ns-options' version?

kellyredding committed:
basic boolean conversion and interpretation behavior

The idea here is to interpret incoming values as booleans in a friendly,
flexible way and be able to easily compare them to other booleans (both
`MuchBoolean` and `TrueClass` and `FalseClass`).  Use `convert` to just
convert a given value to an actual boolean value using much-boolean's
"friendly" interpretation logic.  Create instances using `new` to track
a given values interpreted boolean value and use as a boolean in
comparisons, etc.

Note: this was extracted from the ns-options gem with a few differences:

* no `actual=` writer (this was only needed b/c of ns-options behavior)
* no `returned_value` method (this was only needed b/c of ns-options
  behavior)
* support for "yes/no" type values
* support for additional "true/false" type values
* exposed the underlying conversion logic in a `convert` singleton so
  you don't *have* to build an instance and get its `actual` value
  if all you care about is the actual boolean value

jcredding commented on README.md:
```diff
@@ -4,7 +4,69 @@ An API for friendly boolean parsing and handling
+require 'much-boolean'
+MuchBoolean::FALSE_VALUES # => [nil, 0, "0",
+                                false, "false", "False", "FALSE", "f", "F",
+                                "no", "No", "NO", "n", "N"]
```
@kellyredding - Not sure if you want to put comments in front of these? I'm not sure what's the preferred way to do multi-line return values like this. Anyway, just noticed the syntax wasn't highlighting it as a comment.

jcredding commented on README.md:
```diff
@@ -4,7 +4,69 @@ An API for friendly boolean parsing and handling
+MuchBoolean.parse(0)   # => false
+MuchBoolean.parse('0') # => false
+MuchBoolean.parse(1)   # => false
+MuchBoolean.parse('1') # => false
```
@kellyredding - I think you want these last 2 (the 1's) to be true instead of false.

jcredding commented:
@kellyredding - Looks good :boom:

kellyredding commented:
@jcredding I've updated the readme - good catches.

In doing so, I realized that now that this is no longer in ns-options, the instance has really diminished in value.  I want to keep it for API backwards compatibility but I wanted to add more value too.  So I added a `given` reader so instances can track a given value and its actual boolean value.  Let me know what you think.

kellyredding commented:
@jcredding a few more notes:
- I cleaned up the long array `FALSE_VALUES` output in the README
- I renamed the `parse` method to `convert` as it seems more appropriate
- I updated the "parse" references in descriptions

Let me know what you think of these changes.  Thanks man.

jcredding commented:
@kellyredding - I like convert over parse, it feels better. I'm cool with the `given` logic, do you have a specific use for it? You are right that out of ns-options, the instance doesn't have much value. I'm good with the changes :boom:

kellyredding commented:
@jcredding no I don't have a specific use-case for it right now - just noticed that its only utility was backwards-compatibility and I thought "well, at least with this `given` reader you can track the original value that produced this boolean value."
